geopandas/sindex.py

```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+            please use PyGEOS.
+
+            Parameters
+            ----------
+            geometry : shapely geometry
+                A single shapely geometry to query against the spatial index.
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
+                If predicate is provided, a prepared version of the input geometry is tested using
+                the predicate function against each item in the index whose extent intersects the
+                envelope of the input geometry: predicate(geometry, tree_geometry).
+            sort : bool, default False
+                If True, the results will be sorted in ascending order. If False, results are
+                often sorted but there is no guarantee.
+
+            Returns
+            -------
+            matches : ndarray of shape (n_results, )
+                Integer indices for matching geometries from the spatial index.
+            """  # noqa: E501
+            # handle invalid predicates
+            if predicate not in self.valid_query_predicates:
+                raise ValueError(
+                    "Got `predicate` = `{}`, `predicate` must be one of {}".format(
+                        predicate, self.valid_query_predicates
+                    )
+                )
+            # handle empty / invalid geometries
+            if geometry is None:
+                # this is the behavior in pygeos.strtree.query, we mimic it here
+                return np.array([], dtype=np.intp)
+            if not isinstance(geometry, BaseGeometry):
+                raise TypeError(
+                    "Got `geometry` of type `{}`, `geometry` must be ".format(
+                        type(geometry)
+                    )
+                    + "a shapely geometry."
+                )
+            if geometry.is_empty:
+                return np.array([], dtype=np.intp)
+
+            # get bounds
+            bounds = geometry.bounds  # rtree operates on bounds
+            tree_query = list(self.intersection(bounds, objects=False))
+
+            if not tree_query:
+                return np.array([], dtype=np.intp)
+
+            # check predicate
+            if predicate in ("intersects", "within"):
+                # only contains and intersects are supported by
+                # prepared geometries, see note below regarding within
+                res = []
+                if predicate == "within":
+                    # since these are inverse, we can flip the operation
+                    # and test with prepared predicates from tree
+                    predicate = "contains"
+                for i in tree_query:
+                    if self._prepared_geometries[i] is None:
+                        # if not already prepared, prepare and cache
+                        self._prepared_geometries[i] = prep(self._geometries[i])
+                    if getattr(self._prepared_geometries[i], predicate)(geometry):
```

Comment:
Use of `getattr` masks the operation a bit, and use of "intersects" here (left side tree geom rather than input geom) is not immediately obvious without further explanation.  

Recommend adding a comment like
```
# Test the prepared tree geometry against the input geometry.
# For "intersects" this leverages the cache of prepared tree geometries.
```

Reply:
I've added a lot more comments to this section from the feedback. Please let me know if it seems more or less clear.